### DISCUSSION TITLE
Can we add multiple signatures?

### DISCUSSION THREAD

**Discussion by isnunasrudin** [lsnepomuceno/laravel-a1-pdf-sign#113] [Q&A]
Can we add multiple signatures?

**Comment by lsnepomuceno** [comment]
Hello @isnunasrudin ,

I'm working on version 2.x of the package, and along with it I will bring major improvements, including the possibility of multiple signatures on the same document.

The dev branch has already been created and I will be uploading updates soon.

Branch link: https://github.com/lsnepomuceno/laravel-a1-pdf-sign/tree/v2.x-dev

**Reply by igomis** [reply]
Great Idea !!

**Comment by lsnepomuceno** [comment]
Hello @isnunasrudin,

Following up on this properly: **yes, it shipped in 2.0.**

```php
use LSNepomuceno\LaravelA1PdfSign\Facades\A1PdfSign;

$first = A1PdfSign::newSignature()
    ->certificate('first.pfx', $password)
    ->pdf('contract.pdf')
    ->info(name: 'First signer')
    ->sign();

$path = $first->save(storage_path('contract-signed.pdf'));

$second = A1PdfSign::newSignature()
    ->certificate('second.pfx', $password)
    ->pdf($path)                      // signs the already-signed document
    ->info(name: 'Second signer')
    ->sign();
```

Both signatures stay valid. The reason it did not work before is worth stating, because it was not a missing feature but a wrong approach: 1.x imported every page into a new document and rebuilt it, which discarded annotations, form fields and any signature already present — the behaviour reported in [TCPDF#430](https://github.com/tecnickcom/TCPDF/issues/430).

2.0 **appends a revision** instead (ISO 32000-1 §7.5.6). The original bytes survive byte for byte, so each signature covers the document as it stood when that signature was made, and a reader shows "signature 1 valid, covers revision 1" without invalidating it.

Signature fields must not collide, so each revision gets its own name automatically (`Signature1`, `Signature2`, …). Override it when you need a specific one:

```php
->fieldName('ContractSignature')
```

Verified with poppler's `pdfsig` on a document carrying six signatures — all six report *Signature is Valid*, with correctly progressive coverage. That document ships in [`samples/`](https://github.com/lsnepomuceno/laravel-a1-pdf-sign/tree/main/samples) if you want to inspect it in a reader.

2.0 requires PHP 8.4 and Laravel 13; [UPGRADE.md](https://github.com/lsnepomuceno/laravel-a1-pdf-sign/blob/main/UPGRADE.md) has the migration path.
